src/extension.ts

```diff
@@ -13,273 +13,412 @@
             }
-        });
+            return result;
+        }).sort(
+            (a, b) => {
+                return a.length - b.length;
+            }
+        );
+    }
+    return _sortedWorkspaceFolders || [];
+}
 
-        if (CONFIGURATION.logToFile) {
-            const logPath = workspace.rootPath + '/rls' + Date.now() + '.log';
-            const logStream = fs.createWriteStream(logPath, { flags: 'w+' });
-            logStream.on('open', function (_f) {
-                childProcess.stderr.addListener('data', function (chunk) {
-                    logStream.write(chunk.toString());
-                });
-            }).on('error', function (err: any) {
-                console.error("Couldn't write to " + logPath + ' (' + err + ')');
-                logStream.end();
-            });
+function getOuterMostWorkspaceFolder(folder: WorkspaceFolder): WorkspaceFolder {
+    const sorted = sortedWorkspaceFolders();
+    for (const element of sorted) {
+        let uri = folder.uri.toString();
+        if (uri.charAt(uri.length - 1) !== '/') {
+            uri = uri + '/';
+        }
+        if (uri.startsWith(element)) {
+            return workspace.getWorkspaceFolder(Uri.parse(element)) || folder;
         }
-
-        return childProcess;
-    } catch (e) {
-        stopSpinner('RLS could not be started');
-        throw new Error('Error starting up rls.');
     }
+    return folder;
 }
 
-let lc: LanguageClient;
+function didChangeWorkspaceFolders(e: WorkspaceFoldersChangeEvent, context: ExtensionContext): void {
+    _sortedWorkspaceFolders = undefined;
 
-export async function activate(context: ExtensionContext) {
-    const promise = startLanguageClient(context);
-    configureLanguage(context);
-    registerCommands(context);
-    activateTaskProvider();
-    await promise;
+    for (let folder of e.added) {
+        folder = getOuterMostWorkspaceFolder(folder);
+        if (workspaces.has(folder.uri.toString())) {
+            continue;
+        }
+        for (const f of fs.readdirSync(folder.uri.fsPath)) {
+            if (f === 'Cargo.toml') {
+                const workspace = new Workspace(folder);
+                workspace.start(context);
+                break;
+            }
+        }
+    }
+
+    for (const folder of e.removed) {
+        const ws = workspaces.get(folder.uri.toString());
+        if (ws) {
+            workspaces.delete(folder.uri.toString());
+            ws.stop();
+        }
+    }
 }
 
-async function startLanguageClient(context: ExtensionContext) {
-    if (workspace.rootPath === undefined) {
-        window.showWarningMessage('Startup error: the RLS can only operate on a folder, not a single file');
-        return;
+const workspaces: Map<string, Workspace> = new Map();
+
+// We run one RLS and one corresponding language client per workspace folder
+// (VSCode workspace, not Cargo workspace). This class contains all the per-client
+// and per-workspace stuff.
+class Workspace {
```

Comment:
Hmm, yeah, agree `Workspace` is a bit hard to understand. I've gone for `ClientWorkspace` I don't think it is much better, but I think it is important to note that this is about the language client, rather than the server and about vscode workspaces.